### PR TITLE
Add `{heir, immortal}` option to ets

### PR DESCRIPTION
**Pull request by v0idpwn**
Some times I wanted to have an ets decoupled from any process lifecycle. This PR introduces the `{heir, immortal}` option,
which allows exactly that.

Tables that have the atom `immortal` as a heir don't get deleted upon process termination, nor transferred to another process. Instead, they continue to exist with `owner` = `none`.

If this were to be merged, there are a few necessary tweaks:
- disallowing the ability of setting `heir` for an `ets` with owner `none`
- disallowing `{heir, immortal}` for private tables
- typespec changes 
- extending documentation

My implementation is also probably missing some considerations (it feels __too simple__), but in my tests it work.

### CODE REVIEW

**Comment by essen** [comment]
Ets tables are necessarily tied to an application, because that's where the code using them runs in. Unless you're doing something very special that I'd love to hear about, your `immortal` option isn't better than just creating the ets table in the `AppMod:start/2` function [like Ranch does](https://github.com/ninenines/ranch/blob/master/src/ranch_app.erl#L25-L26).

Could you explain the use case and why you really need an owner-free ets table?

Note that I'm not part of OTP team, but I believe these kinds of proposals have better luck submitted as an EEP with a detailed rationale behind the changes.

**Comment by juhlig** [comment]
This sounds dubious to me too, and I would be interested in hearing a use case.

FWIW, you can achieve basically the same thing right now by putting the kernel sup as heir.

**Comment by v0idpwn** [comment]
Thanks for your questions.

For context, my company has a library that performs some start-up:
- Starts atomics
- Starts an ETS table
- Sets some keys persistent_term

Of these three resources, the only one that has the lifecycle related to the starting process is the ETS table. When the caller (in this case, the dependent application, and not mine) is restarted, the startup function is called again, and what it finds is that the persistent_term keys are there, the atomics are there, but the ETS table is removed. In my case, there's some benefit if the ETS is kept between restarts. For this reason, I thought it might be a good idea to allow the table to persist in such situations. 

There are different ways to work around the above issue, I have considered some and implemented a solution. But I think an option to make it "native" would be aligned with other APIs and provide a better experience than, e.g.: setting `kernel_sup` as a heir or changing the process architecture to avoid the ETS termination.

In short, I don't think it's strictly necessary to add this option, but that it would be more consistent with other APIs and a nice to have.

With regards to the need for an EEP, I thought this change was too small in scope and thus, undeserving of one. I was planning to open an issue, but decided to give a stab at the implementation instead :)

**Comment by max-au** [comment]
> Of these three resources, the only one that has the lifecycle related to the starting process is the ETS table. 

I always considered lack of ownership concept for `persistent_term` as an implementation flaw. It breaks one of the most important abstractions, - supervision. Having state that survives owner process restart defeats the purpose of having supervision tree, because restarted processes still have the same (broken) state. That said, I'd rather file a bug report/proposal to bind every persistent term to the owning process.

The use-case, and library, you explained, reminds me very much of my own https://github.com/max-au/ecount But it has no problem with "dependent application restart", because it binds lifecycle of the persistent term it uses to the `gen_server` lifecycle. In other words, if you restart your "dependent application", there are no "startup function" it calls. The problem simply does not exist. Would you consider that solution instead?

**Comment by Maria-12648430** [comment]
> * disallowing `{heir, immortal}` for private tables

Shouldn't that also apply for `protected` tables? Because otherwise, the table can only be read but not changed or deleted.

Also, immortality should only be possible for _named_ tables, because otherwise how are you going to find back your table?

That said, there _may_ be use cases for immortal tables, I don't know, but I don't think it should be encouraged. I believe that a implementation that relies on this behavior (ie, distinct applications relying on and directly accessing the same table) is at least questionable. A table _should_ have a process responsible for it.

**Comment by max-au** [comment]
>  should the lifecycle of persistent terms be "linked" with the writing processes?

Haven't thought of it that way. Not sure about the writer, should it be the first, or the last one?

**Comment by v0idpwn** [comment]
I think the latest one would make more sense. Were you considering a different approach?
